FWCore/Integration/plugins/SourceWithWaits.cc

```diff
@@ -63,98 +63,220 @@ namespace edmtest {
         eventsPerLumi_(pset.getUntrackedParameter<std::vector<unsigned int>>("eventsPerLumi")),
-        lumisPerRun_(pset.getUntrackedParameter<unsigned int>("lumisPerRun")) {}
+        lumisPerRun_(pset.getUntrackedParameter<unsigned int>("lumisPerRun")),
+        multipleEntriesForRun_(pset.getUntrackedParameter<unsigned int>("multipleEntriesForRun")),
+        multipleEntriesForLumi_(pset.getUntrackedParameter<unsigned int>("multipleEntriesForLumi")),
+        declareLast_(pset.getUntrackedParameter<bool>("declareLast")),
+        declareAllLast_(pset.getUntrackedParameter<bool>("declareAllLast")) {}
 
   SourceWithWaits::~SourceWithWaits() {}
 
   void SourceWithWaits::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
     edm::ParameterSetDescription desc;
-    desc.addUntracked<unsigned int>("timePerLumi");
+    desc.addUntracked<double>("timePerLumi");
+    desc.addUntracked<double>("sleepAfterStartOfRun");
     desc.addUntracked<std::vector<unsigned int>>("eventsPerLumi");
     desc.addUntracked<unsigned int>("lumisPerRun");
+    desc.addUntracked<unsigned int>("multipleEntriesForRun", 0);
+    desc.addUntracked<unsigned int>("multipleEntriesForLumi", 0);
+    desc.addUntracked<bool>("declareLast", false);
+    desc.addUntracked<bool>("declareAllLast", false);
     descriptions.add("source", desc);
   }
 
-  edm::InputSource::ItemType SourceWithWaits::getNextItemType() {
+  edm::InputSource::ItemTypeInfo SourceWithWaits::getNextItemType() {
     constexpr unsigned int secondsToMicroseconds = 1000000;
 
+    if (startedNewRun_) {
+      usleep(secondsToMicroseconds * sleepAfterStartOfRun_);
+      startedNewRun_ = false;
+    }
+
+    if (lastEventOfLumi_ || noEventsInLumi_) {
+      usleep(secondsToMicroseconds * timePerLumi_ / (eventsPerLumi_[currentLumi_ - 1] + 1));
+      lastEventOfLumi_ = false;
+      noEventsInLumi_ = false;
+    }
+
     // First three cases are for the initial file, run, and lumi transitions
     // Note that there will always be at exactly one file and at least
     // one run from this test source.
     if (currentFile_ == 0u) {
       ++currentFile_;
-      return edm::InputSource::IsFile;
-    } else if (currentRun_ == 0u) {
+      return edm::InputSource::ItemType::IsFile;
+    }
+    // First Run
+    else if (currentRun_ == 0u) {
       ++currentRun_;
-      return edm::InputSource::IsRun;
-    } else if (currentLumi_ == 0u) {
+      if (currentRun_ != multipleEntriesForRun_) {
+        startedNewRun_ = true;
+        return (declareLast_ || declareAllLast_)
+                   ? edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun,
+                                                    edm::InputSource::ItemPosition::LastItemToBeMerged)
+                   : edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun,
+                                                    edm::InputSource::ItemPosition::NotLastItemToBeMerged);
```

Comment:
I suppose this could be also
```suggestion
        return edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun,
                                                    (declareLast_ || declareAllLast_) ? edm::InputSource::ItemPosition::LastItemToBeMerged : edm::InputSource::ItemPosition::NotLastItemToBeMerged);
```
or
```suggestion
        auto const position = (declareLast_ || declareAllLast_) ? edm::InputSource::ItemPosition::LastItemToBeMerged : edm::InputSource::ItemPosition::NotLastItemToBeMerged);
        return edm::InputSource::ItemTypeInfo(edm::InputSource::ItemType::IsRun, position);
```

I'm not really sure which one to prefer though.